README.md
# react-start

这个项目基础搭建终于完成，用于学习webpack+eslint+babel+react相关的知识点。
项目配还是很清晰的，后续接着完善这个项目。

## 功能实现

通过npm start 启动项目,项目支持热更新,具体支持的功能看webpack.config.js文件。里头写还是比较详细的。


## 全局环境 

```shell
Node >= 8.x
NPM >= 5.x
```

### step 1:
`npm run dll`

### step 2:
`npm start` or `npm run start`


CHANGELOG.md
### 项目搭建

2018.07.17 
项目基础搭建完成。下个版本还需要加入rxjs,router,svg字体等。

src/index.js
import _ from 'lodash';
import printMe from './print.js';
import './styles.css'


function component() {
    var element = document.createElement('div');
    var btn = document.createElement('button');
    element.innerHTML = _.join(['Hello', 'webpack'], ' ');
    btn.innerHTML = 'Click me and check the console!';
    btn.onclick = printMe;
    element.appendChild(btn);
    return element;
}

let element = component(); // 当 print.js 改变导致页面重新渲染时，重新获取渲染的元素
document.body.appendChild(element);

if (module.hot) {
    module.hot.accept('./print.js', function () {
        console.log('Accepting the updated printMe module!');
        document.body.removeChild(element);
        element = component(); // 重新渲染页面后，component 更新 click 事件处理
        document.body.appendChild(element);
        // 社区还有许多其他 loader 和示例，可以使 HMR 与各种框架和库(library)平滑地进行交互……
        // 引入这些loader就可以不用手动的去写上面的操作了。
        // 例如Vue Loader，Angular HMR，React Hot Loader
    })
}

webpack.dev.js
const path = require('path');
const webpack = require('webpack');
const ROOT_PATH = path.resolve(__dirname);
const APP_PATH = path.resolve(ROOT_PATH, 'src');
const BUILD_PATH = path.resolve(ROOT_PATH, 'build');

module.exports = {
  devtool: 'cheap-module-eval-source-map',
  mode: 'development',
  entry: ['./src/app.jsx', 'react-hot-loader/patch'],
  devServer: {
    // 以下配置告知 webpack-dev-server，在 localhost:3000 下建立服务，将 dist 目录下的文件，作为可访问文件。
    publicPath: '/',  // 默认 publicPath 是 "/"，所以你的包(bundle)可以通过 http://localhost:3000/bundle.js 访问。
    contentBase: BUILD_PATH,
    open: true,
    openPage: '', // 指定在打开浏览器时导航到的页面。
    hot: true, // 热更新
    port: 3000,
    host: "localhost", // 如果你希望服务器外部可访问
    compress: true, // 开发服务器是否启动gzip等压缩
    inline: true, // 应用程序启用内联模式(inline mode)。这意味着一段处理实时重载的脚本被插入到你的包(bundle)中
    quiet: false, // 启用 quiet 后，除了初始启动信息之外的任何内容都不会被打印到控制台。这也意味着来自 webpack 的错误或警告在控制台不可见。
    proxy: {}// proxy URLs to backend development server
},
  plugins: [
    new webpack.NamedModulesPlugin(),
    new webpack.HotModuleReplacementPlugin(),
  ],
  watch: true, // 这意味着在初始构建之后，webpack 将继续监听任何已解析文件的更改。Watch 模式默认关闭。
  watchOptions: {
      ignored: /node_modules/, // 忽略不用监听变更的目录，监听大量文件系统会导致大量的 CPU 或内存占用。
      aggregateTimeout: 500, // 防止重复保存频繁重新编译,500毫米内重复保存不打包
      poll: 1000 // 多少毫秒检查一次变动 指定毫秒为单位进行轮询
  }
};


webpack.config.js
const path = require('path');
const webpack = require('webpack');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const AddAssetHtmlPlugin = require('add-asset-html-webpack-plugin');
const MiniCssExtractPlugin = require('mini-css-extract-plugin');
const WebpackVersionPlugin = require('webpack-version-plugin');
const WebpackMerge = require('webpack-merge');
const ROOT_PATH = path.resolve(__dirname);
const APP_PATH = path.resolve(ROOT_PATH, 'src');
const BUILD_PATH = path.resolve(ROOT_PATH, 'build');
const manifest = require('./vendor/react.manifest.json');
const svgDirs = [
  path.resolve(ROOT_PATH, 'src/assets/svg') // 自己私人的 svg 存放目录
];

module.exports = WebpackMerge(
  process.env.NODE_ENV === 'dev' ? require('./webpack.dev.js') : require('./webpack.build.js'),
  {
    output: {
      path: BUILD_PATH, // 编译到当前目录
      filename: 'js/[name].js'
    },
    module: {
      rules: [
        {
          test: /\.css$/,
          include: [path.resolve('src')],
          use: ['css-hot-loader', 'style-loader', 'css-loader']
        },
        {
          test: /\.scss$/,
          use: [
            'css-hot-loader',
            MiniCssExtractPlugin.loader,
            'css-loader?importLoaders=1', // 在 css-loader 前应用的 loader 的数量
            'sass-loader'
          ]
        },
        {
          test: /\.jsx?$/, // 用babel编译jsx和es6
          include: [APP_PATH], // 指定检查的目录
          exclude: /node_modules/,
          use: [
            'cache-loader', // 在一些性能开销较大的 loader 之前添加此 loader，以将结果缓存到磁盘里。
            'thread-loader', // 把这个 loader 放置在其他 loader 之前， 放置在这个 loader 之后的 loader 就会在一个单独的 worker 池(worker pool)中运行
            {
              loader: 'babel-loader',
              options: {
                cacheDirectory: true, //使用默认缓存目录，尝试读取缓存，来避免在每次执行时，可能产生的、高性能消耗的 Babel 重新编译过程
                // modules关闭 Babel 的模块转换功能，保留原本的 ES6 模块化语法
                presets: ['react', 'stage-2', ['env', { modules: false }]],
                plugins: [
                  'transform-runtime',
                  'transform-decorators-legacy',
                  'transform-class-properties',
                  'react-hot-loader/babel'
                ]
              }
            }
          ]
        },
        {
          test: /\.(woff|woff2|eot|ttf)(\?.*$|$)/,
          use: ['url-loader'],
          exclude: [svgDirs]
        },
        {
          test: /\.(png|jpg|gif)$/,
          use: ['url-loader?limit=8192&name=images/[hash:8].[name].[ext]']
        },
        {
          test: /\.(svg)$/i,
          use: ['svg-sprite-loader'],
          include: svgDirs // 把 svgDirs 路径下的所有 svg 文件交给 svg-sprite-loader 插件处理
        }
      ]
    },
    plugins: [
      // HtmlWebpackPlugin 创建了一个全新的文件，所有的 bundle 会自动添加到 html 中。
      new HtmlWebpackPlugin({
        filename: './index.html', // 生成的html存放路径，相对于 path
        template: './src/index.html',
        inject: true, // 允许插件修改哪些内容，包括head与body
        hash: true // 为静态资源生成hash值
      }),
      // 该插件将给定的JS或CSS文件添加到Webpack知道的文件中，并将其放入资源列表中，html-webpack-plugin注入生成的html。
      new AddAssetHtmlPlugin({
        filepath: path.resolve(ROOT_PATH, 'vendor/*.dll.js')
      }),
      // 在这引用之前打包出来的依赖
      new webpack.DllReferencePlugin({
        context: ROOT_PATH,
        manifest,
        extensions: ['.js', '.jsx']
      }),
      /**
       * 此插件将CSS提取到单独的文件中。它为每个包含CSS的JS文件创建一个CSS文件。
       * 它支持CSS和SourceMaps的按需加载。
       **/
      new MiniCssExtractPlugin({
        filename: 'css/[name].css'
      }),

      new WebpackVersionPlugin({
        cb: hashMap => {
          console.log(hashMap);
        }
      })
    ],
    resolve: {
      modules: ['node_modules', path.join(ROOT_PATH, './node_modules')],
      extensions: ['.js', '.jsx', '.scss', '.json'],
      alias: {
        components: path.resolve(APP_PATH, './components')
      }
    }
  }
);


webpack.dll.config.js
const path = require('path');
const webpack = require('webpack');
const CleanWebpackPlugin = require('clean-webpack-plugin');
const WebpackVersionPlugin = require('webpack-version-plugin');

module.exports = {
  entry: {
    react: ['react', 'react-dom', 'react-redux'] // react模块打包到一个动态连接库
  },
  output: {
    path: path.resolve(__dirname, 'vendor/'), 
    filename: '[name].dll.js', // 输出动态连接库的文件名称  
    library: '_dll_[name]_[hash]' // 全局变量名称
  },
  devtool: '#source-map',
  plugins: [
    new webpack.DllPlugin({
      context: __dirname,
      name: '_dll_[name]_[hash]', // 和output.library中一致，也就是输出的manifest.json中的 name值
      path: path.join(__dirname, 'vendor', '[name].manifest.json') // [name].manifest.json 的文件，这个文件是用来让 DLLReferencePlugin 映射到相关的依赖上去的
    }),
    new CleanWebpackPlugin(['vendor'], { root: path.resolve() }), // 清空文件
    new WebpackVersionPlugin({
      cb: (hashMap) => {
        console.log(hashMap);
      }
    })
  ]
};


webpack.all.js
const path = require('path');
const webpack = require('webpack');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const AddAssetHtmlPlugin = require('add-asset-html-webpack-plugin');
const ForkTsCheckerWebpackPlugin = require('fork-ts-checker-webpack-plugin');
const ExtractTextPlugin = require('extract-text-webpack-plugin');
const MiniCssExtractPlugin = require('mini-css-extract-plugin');
const UglifyJsPlugin = require('uglifyjs-webpack-plugin');
const CleanPlugin = require('clean-webpack-plugin');
const OptimizeCssAssetsPlugin = require('optimize-css-assets-webpack-plugin');
const PostcssPxtorem = require('postcss-pxtorem');
const AutoPrefixer = require('autoprefixer'); // eslint-disable-line
const WebpackVersionPlugin = require('webpack-version-plugin');
const StyleLintPlugin = require('stylelint-webpack-plugin');

const ROOT_PATH = path.resolve(__dirname);
const APP_PATH = path.resolve(ROOT_PATH, 'src');
const BUILD_PATH = path.resolve(ROOT_PATH, 'build');
const svgDirs = [
  path.resolve(__dirname, 'src/assets/svg') // 自己私人的 svg 存放目录
];

const env = process.env.NODE_ENV;
const manifest = require('./vendor/react.manifest.json');
const { theme } = require('./package.json');

module.exports = {
  devtool: env === 'production' ? false : 'cheap-module-eval-source-map',
  entry: { app: ['./src/index.jsx', 'react-hot-loader/patch'] },
  mode: env === 'production' ? 'production' : 'development',
  output: {
    path: BUILD_PATH, // 编译到当前目录
    filename: 'js/[name].js'
  },
  devServer: {
    publicPath: '/',
    contentBase: BUILD_PATH,
    historyApiFallback: true,
    hot: true,
    open: true,
    inline: true,
    port: 8888,
    host: 'localhost',
    openPage: '',
    proxy: {},
    quiet: true,
    compress: true // 开发服务器是否启动gzip等压缩
    /*  https: {
      key: fs.readFileSync('/path/to/server.key'),
      cert: fs.readFileSync('/path/to/server.crt'),
      ca: fs.readFileSync('/path/to/ca.pem')
    } */
  },
  module: {
    rules: [
      {
        test: /\.scss$/,
        // css-hot-loader会增加打包的体积
        use:
          env === 'production'
            ? [
              MiniCssExtractPlugin.loader,
              'css-loader?importLoaders=1',
              'postcss-loader',
              'sass-loader'

            ]
            : [
              'css-hot-loader',
              MiniCssExtractPlugin.loader,
              'css-loader?importLoaders=1',
              'postcss-loader',
              'sass-loader'
            ]
        /*  use: ExtractTextPlugin.extract({
          fallback: 'style-loader',
          use: ['css-hot-loader', 'css-loader', 'postcss-loader', 'sass-loader']
        }) */
      },
      {
        test: /\.less$/,
        use: [
          'css-hot-loader',
          { loader: 'style-loader', options: { sourceMap: true } },
          { loader: 'css-loader', options: { sourceMap: true } },
          {
            loader: 'postcss-loader',
            options: {
              ident: 'postcss',
              plugins: loader => [
                PostcssPxtorem({
                  rootValue: 100,
                  propWhiteList: ['*']
                }),
                AutoPrefixer({
                  browsers: ['last 2 versions', '> 1%', 'ie >= 8']
                })
              ]
            }
          },
          { loader: 'less-loader', options: { sourceMap: true, modifyVars: theme } }
        ]
      },
      {
        test: /\.css$/,
        include: [path.resolve('src')],
        use: ['css-hot-loader', 'style-loader', 'css-loader', 'postcss-loader']
      },
      {
        test: /\.jsx?$/,
        enforce: 'pre',
        use: [
          {
            loader: 'eslint-loader',
            options: { fix: true }
          }
        ],
        include: [path.resolve(__dirname, 'src')], // 指定检查的目录
        exclude: /node_modules/
      },

      {
        test: /\.jsx?$/, // 用babel编译jsx和es6
        include: [path.resolve(__dirname, 'src')], // 指定检查的目录
        exclude: /node_modules/,
        loader: 'babel-loader',
        options: {
          cacheDirectory: true,
          presets: ['react', 'stage-2', ['env', { modules: false }]],
          // modules关闭 Babel 的模块转换功能，保留原本的 ES6 模块化语法
          plugins: ['transform-runtime', 'transform-decorators-legacy', 'react-hot-loader/babel']
        }
      },
      {
        test: /\.(woff|woff2|eot|ttf)(\?.*$|$)/,
        use: ['url-loader']
      },
      {
        test: /\.(svg)$/i,
        use: ['svg-sprite-loader'],
        include: svgDirs // 把 svgDirs 路径下的所有 svg 文件交给 svg-sprite-loader 插件处理
      },
      {
        test: /\.(png|jpg)$/,
        use: ['url-loader?limit=8192&name=images/[hash:8].[name].[ext]']
      }
    ]
  },
  plugins: [
    new HtmlWebpackPlugin({
      filename: './index.html', // 生成的html存放路径，相对于 path
      template: './src/index.html',
      inject: true, // 允许插件修改哪些内容，包括head与body
      hash: true // 为静态资源生成hash值
    }),
    new AddAssetHtmlPlugin({
      filepath: path.resolve(__dirname, 'build/vendor/*.dll.js')
    }),
    new ExtractTextPlugin({
      // 指定css文件名 打包成一个css
      filename: 'css/[name].css',
      disable: false,
      allChunks: true
    }),
    // 在这引用之前打包出来的依赖
    new webpack.DllReferencePlugin({
      context: __dirname,
      manifest,
      extensions: ['.js', '.jsx']
    }),
    new MiniCssExtractPlugin({
      filename: 'css/[name].css'
    }),
    new StyleLintPlugin(),
    new ForkTsCheckerWebpackPlugin({
      checkSyntacticErrors: true,
      tsconfig: path.resolve(__dirname, 'tsconfig.json'),
      tslint: path.resolve(__dirname, 'tslint.json'),
      watch: ['./src/**/*.tsx'],
      ignoreLints: ['no-console', 'object-literal-sort-keys', 'quotemark']
    }),
    new webpack.NamedModulesPlugin(),
    new webpack.HotModuleReplacementPlugin(),
    new webpack.WatchIgnorePlugin([/\.js$/]), // 忽略掉 d.ts 文件，避免因为编译生成 d.ts 文件导致又重新检查。
    new webpack.optimize.ModuleConcatenationPlugin(),
    // 3.0新功能 范围提升 （Scope Hoisting ）,作用域提升，这是在webpack3中所提出来的。它会使代码体积更小，因为函数申明语句会产生大量代码.
    new WebpackVersionPlugin({
      cb: (hashMap) => {
        console.log(hashMap);
      }
    })
  ],

  resolve: {
    modules: ['node_modules', path.join(__dirname, './node_modules')],
    extensions: ['.js', '.jsx', '.less', '.scss', '.json'],
    alias: {
      components: path.resolve(APP_PATH, './components')
    }
  },
  // 防止将某些 import 的包(package)打包到 bundle 中，而是在运行时(runtime)再去从外部获取这些扩展依赖 例如，从 CDN 引入 jQuery，而不是把它打包.
  externals: {
  },
  watch: env === 'development',
  watchOptions: {
    ignored: /node_modules/, // 忽略不用监听变更的目录
    aggregateTimeout: 500, // 防止重复保存频繁重新编译,500毫米内重复保存不打包
    poll: 1000 // 每秒询问的文件变更的次数
  }
};

if (env === 'production') {
  module.exports.optimization = {
    minimizer: [
      new UglifyJsPlugin({
        cache: true,
        parallel: true,
        uglifyOptions: {
          compress: true,
          ecma: 6,
          mangle: true
        },
        sourceMap: true
      })
    ]
    // 这个变化还是很大的，之前的webpack版本用的都是commonchunkplugin，但是webpack4开始使用common-chunk-and-vendor-chunk 配置如下:
    /* splitChunks: {
      cacheGroups: {
        commons: {
          chunks: 'initial',
          minChunks: 2,
          maxInitialRequests: 5, // The default limit is too small to showcase the effect
          minSize: 0 // This is example is too small to create commons chunks
        },
        vendor: {
          test: /node_modules/,
          chunks: 'initial',
          name: 'vendor',
          priority: 10,
          enforce: true
        }
      }
    } */
  };

  module.exports.plugins = (module.exports.plugins || []).concat([
    new OptimizeCssAssetsPlugin({})
    // new CleanPlugin([BUILD_PATH])
  ]);
}
